### PR TITLE
Fix thread safety in async code

### PR DESCRIPTION
**Pull request by valentiniliescu**
#70

### CODE REVIEW

File: Refactoring.Pipelines/JoinedPipes.cs [hunk 1]
```diff
@@ -21,8 +22,8 @@ public class JoinedPipes<TInput1, TInput2> : Sender<Tuple<TInput1, TInput2>>, IG
         private readonly ForwardingListener<TInput2> _listener2;
         private readonly ISender<TInput1> _sender1;
         private readonly ISender<TInput2> _sender2;
-        private readonly Queue<TInput1> _values1 = new Queue<TInput1>();
-        private readonly Queue<TInput2> _values2 = new Queue<TInput2>();
+        private readonly ConcurrentQueue<TInput1> _values1 = new ConcurrentQueue<TInput1>();
+        private readonly ConcurrentQueue<TInput2> _values2 = new ConcurrentQueue<TInput2>();
```
Comment (valentiniliescu) [inline Refactoring.Pipelines/JoinedPipes.cs:26]: Not sure if I need to subclass and create a new JoinedPipes in the Async namespace with just the ConcurrentQueue part.

**Comment by valentiniliescu** [comment]
@JayBazuzi @isidore is there a way to run the flaky test multiple times in the CI pipeline, to see if the issue is fixed? I ran it and it passed multiple times locally., and I was able to reproduced the issue pretty consistently before the fix.

File: Refactoring.Pipelines/JoinedPipes.cs [hunk 2]
```diff
@@ -63,7 +64,9 @@ private void _SendIfReady()
         {
             if (_values1.Any() && _values2.Any())
             {
-                _Send(Tuple.Create(_values1.Dequeue(), _values2.Dequeue()));
+                _values1.TryDequeue(out TInput1 value1);
```
Comment (JayBazuzi) [inline Refactoring.Pipelines/JoinedPipes.cs:67]: Can/should this use `out var`?

**Review by JayBazuzi — COMMENTED** [review]
This is probably not the last time we write this kind of defect. What is something we can do to reduce the chance of this happening again?

File: Refactoring.Pipelines/JoinedPipes.cs (hunk 2 above)
Comment (valentiniliescu) [reply Refactoring.Pipelines/JoinedPipes.cs:67]: My personal opinion is to use "var" only when the type is obvious (e.g. var something = new MyClass()), but I don't feel strongly about that. Is there a ReSharper settings file for the project that I can use to format the code? Or maybe there is, but I don't know how to apply it.

**Comment by valentiniliescu** [comment]
> This is probably not the last time we write this kind of defect. What is something we can do to reduce the chance of this happening again?

Checking/ensuring thread-safety is a hard problem. Let's have a video chat about that.

**Pull request merged**